src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs

```diff
@@ -10,6 +10,69 @@ namespace System.IO
+        /// <param name="fileInfo">The object describing a file that does not exist in disk yet.</param>
+        /// <param name="mode">A constant that determines how to open or create the file.</param>
+        /// <param name="rights">A constant that determines the access rights to use when creating access and audit rules.</param>
+        /// <param name="share">A constant that determines how the file will be shared by processes.</param>
+        /// <param name="bufferSize">A positive <see cref="int" /> value greater than 0 indicating the buffer size. The default buffer size is 4096.</param>
+        /// <param name="options">A constant that specifies additional file options.</param>
+        /// <param name="fileSecurity">An object that determines the access control and audit security for the file.</param>
+        /// <returns>A file stream that represents the newly created file.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="fileSecurity" /> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="bufferSize" /> is negative or zero.
+        ///
+        /// -or-
+        ///
+        /// <paramref name="mode" />, <paramref name="rights" />, <paramref name="share" /> or <paramref name="options" /> contain an invalid value.
+        /// </exception>
+        /// <remarks><format type="text/markdown"><![CDATA[
+        /// ## Remarks
+        ///
+        /// This method can also throw any of the exceptions that <xref:System.IO.FileStream> can throw.
+        /// ]]></format></remarks>
+        public static FileStream Create(this FileInfo fileInfo, FileMode mode, FileSystemRights rights, FileShare share, int bufferSize, FileOptions options, FileSecurity fileSecurity)
+        {
+            string badArg = null;
+
+            if (mode < FileMode.CreateNew || mode > FileMode.Append)
+                badArg = nameof(mode);
+
+            if (rights < FileSystemRights.ReadData || rights > FileSystemRights.FullControl)
+                badArg = nameof(rights);
+
+            if (share < FileShare.None || share > FileShare.Inheritable)
+                badArg = nameof(share);
+
+            if (options < FileOptions.None || options > FileOptions.Asynchronous)
+                badArg = nameof(options);
+
+            if (badArg != null)
+                throw new ArgumentOutOfRangeException(badArg, SR.ArgumentOutOfRange_Enum);
+
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException(nameof(bufferSize));
+
+            if (fileSecurity == null)
+                throw new ArgumentNullException(nameof(fileSecurity));
+
+            fileInfo.SetAccessControl(fileSecurity);
+            return new FileStream(fileInfo.FullName, mode, mode == FileMode.Append ? FileAccess.Write : FileAccess.ReadWrite, share, bufferSize, options);
```

Comment:
@JeremyKuhne I have some questions about these two lines above:

1. Is calling SetAccessControl (defined in this same file) the correct way to set the access control for the fileSecurity object, before the file actually exists? It will internally call the Persists method, which doesn't seem to care if the file doesn't exist, because it knows it should pass a null SafeHandle in that case ([source](https://github.com/dotnet/corefx/blob/0cc22ccba707d7552159363d586262fa9c4e8a2a/src/System.Security.AccessControl/src/System/Security/AccessControl/NativeObjectSecurity.cs#L349)).

2. The FileAccess value I'm passing to the FileStream constructor is based on the FileMode value that was passed to the Create method. Is that the correct thing to do? We do that in one of the FileStream constructors ([source](https://github.com/dotnet/corefx/blob/ac99a1b7168bd32046a954c3f06012c0fa909bed/src/Common/src/CoreLib/System/IO/FileStream.cs#L166)).

Reply:
This will be addressed in the File Create PR.